File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        """
            0. loop once to find size
            1. loop with curr and prev, sending curr to size-nth node
            2. assign prev.next to curr.next
            3. return head

            Edge case: remove head, return head.next
        """
        if not head:
            return None
        
        size = 0
        curr = head
        while curr:
            curr = curr.next
            size += 1

        if (size - n) == 0:
            return head.next

        prev, curr = None, head
        i = 0
        while curr and i < (size - n):
            prev = curr
            curr = curr.next
            i += 1
        
        prev.next = curr.next
        return head
        

File: Data Structures & Algorithms/reverse-a-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (!head) { return head; }
        
        ListNode* newHead = head;
        if (head->next) {
            newHead = reverseList(head->next);
            head->next->next = head;
        }
        head->next = nullptr;

        return newHead;
    }
};

/*
    0. make a pointer to the head
    1. push every element of the list onto a stack
    2. loop from the back and assign every element to the list
    3. return the new list
*/
